### PR TITLE
Use trait to abstract emit modes

### PR DESCRIPTION
**Pull request by bash**

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -469,10 +474,23 @@ impl<'b, T: Write + 'b> Session<'b, T> {
     }
 }
 
+pub(crate) fn create_emitter<'a>(config: &Config) -> Box<dyn Emitter + 'a> {
+    match config.emit_mode() {
+        EmitMode::Files if config.make_backup() => Box::new(emitter::FilesWithBackupEmitter::new()),
```
Comment (topecongiro) [inline src/lib.rs]: I prefer to use `default()` over an empty constructor.

File: src/test/mod.rs [hunk 2]
```diff
@@ -344,9 +344,9 @@ fn stdin_formatting_smoke_test() {
     }
 
     #[cfg(not(windows))]
-    assert_eq!(buf, "fn main() {}\n".as_bytes());
+    assert_eq!(buf, "stdin:\n\nfn main() {}\n".as_bytes());
     #[cfg(windows)]
-    assert_eq!(buf, "fn main() {}\r\n".as_bytes());
+    assert_eq!(buf, "stdin:\n\r\nfn main() {}\r\n".as_bytes());
```
Comment (topecongiro) [inline src/test/mod.rs]: `writeln!` only emits the line feed character [1], so the first `\r` is unnecessary here.

[1] https://doc.rust-lang.org/std/macro.writeln.html

File: src/lib.rs (hunk 1 above)
Comment (bash) [reply src/lib.rs]: Addressed in dc1a440

File: src/test/mod.rs (hunk 2 above)
Comment (bash) [reply src/test/mod.rs]: That explains why the windows build was failing 😄. Thanks.

Comment (bash) [reply src/test/mod.rs]: Fixed in a759b83

**Comment by topecongiro** [comment]
LGTM, thank you!

**Pull request merged**